Avoid overly-cautious type guards in favor of more performant type discriminators

nabbydude opened the pull request:
**Description**
Adds `Editor.isEditorNode`, `Element.isElementNode`, and `Text.isTextNode` that take a `Node` argument and check quickly if it's a specific node type.

Also fixes classic `isEditor` to actually check for all properties, actually respect deepness settings (a bug I only noticed because I was editing the file 😅), and also respect those deepness settings for its operation list.

This also removes any calls to `Element.isAncestor`, since it was only ever used where `!Text.isTextNode` would function just fine. It doesn't remove the function itself since its user accessible. Rigourously testing for any of two node types from `any` like this is something I cant think of a real use case for and it might be worth deprecating.

**Issue**
Fixes #5980
Basically checking if something is an object all the time and checking it has every single function its supposed to is a waste of performance, 99% of the time we can just quickly check if its definitely not one of the other two types and move on.

**Context**
This touches a lot of files, since it replaces almost every call of  `isEditor`, `isElement`, and `isText` in the `slate` package.
I've removed a few redundant calls and cleaned up a few spaghetti boolean expressions.

`Editor.isEditor` previously only passed its first argument to the function it wrapped, which meant deepness options were ignored. this PR just adds that function directly to the object without wrapping it to avoid that happening again. `isEditorNode` is included in the same way.

I have another open PR (#5970) that changes `normalizeNode`, which this will conflict with. If both end up approved I have a merging commit ready.

**Alternatives**
I chose these names mostly to be close to the originals in alphabetical order, and to keep the `isX` paradigm. other options could be `nodeIsEditor`, `isNodeEditor`, `Node.isEditor`. I can change them if something else is preferred.

We could overwrite the old functionality and migrate the cautious type guard to `verifyElement` or some other name but that might break existing user code.

**Future**
- [ ] Add docs
- [x] Use in other packages in repo
- [x] Use in examples

**Checks**
- [x] The new code matches the existing patterns and styles.
- [x] The tests pass with `yarn test`.
- [x] The linter passes with `yarn lint`. (Fix errors with `yarn fix`.)
- [x] The relevant examples still work. (Run examples with `yarn start`.)
- [x] You've [added a changeset](https://github.com/atlassian/changesets/blob/master/docs/adding-a-changeset.md) if changing functionality. (Add one with `yarn changeset add`.)

12joan commented on packages/slate/src/transforms-node/wrap-nodes.ts:
```diff
@@ -70,16 +71,14 @@ export const wrapNodes: NodeTransforms['wrapNodes'] = (
+          mode: 'lowest',
+          voids,
+        })
+      : [[editor, []] as NodeEntry]
```
I find the prior version of this a little easier to read, even if it's more indirect in `match`ing just the editor. Are you able to write the new version without a type assertion?

12joan commented on packages/slate/src/editor/is-editor.ts:
```diff
@@ -13,26 +13,95 @@ export const isEditor: EditorInterface['isEditor'] = (
   }
 
   const isEditor =
```
I'm not sure whether it's worth including this many `typeof value.x === 'function'` checks. Each one carries a small overhead, and there's a big difference between having 16 of them versus 86 (apologies if I miscounted).

Here's a benchmark I ran to compare these numbers of `typeof` checks. Since the numbers don't mean much in isolation, I also included a [`WeakMap.prototype.set`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap/set) call for comparison. Setting and getting weak map values tends to be relatively slow, so anything slower than it should generally be avoided if possible.

https://jsbm.dev/xM6mIrc116gyW - Longer bars are faster.

<img width="443" height="437" alt="image" src="https://github.com/user-attachments/assets/cce252f9-a963-45a9-aa5d-b5ff6e9fda6e" />

nabbydude replied:
Yeah, my intention would be that `Editor.isEditor` is basically never used. The only current use case I can imagine in the current codebase is *maybe* the sanity check that `slate-react`'s `Slate` component does when its first passed an `editor` prop but really that should be unnecessary in a typescripted world.

As I'm typing though I do see the perspective of any user code that updates its dependency without tweaking its calls getting an unexpected performance hit. 🤔

nabbydude commented on .changeset/rich-cheetahs-battle.md:
```diff
@@ -0,0 +1,7 @@
+---
+
+Add `Editor.isEditorNode`, `Element.isElementNode`, and `Text.isTextNode` as alternative type guards for when we already know the object is a node.
+Use these new functions instead of `isEditor`, `isElement`, and `isText` whenever possible, the classic functions are only necessary for typechecking an entirely unknown object.
```
I definitely think so, user code should only need the old functions for serializing/deserializing etc, basically call match functions etc are better off with the new ones.

Just didn't want to cement them in too deep before getting feedback on the naming. Thankfully old code still functions, just inefficiently (equally as inefficiently as before, to be precise)

nabbydude commented:
> The other packages don't use the live versions; I believe they use the transpiled versions of the packages found in `dist`. There should be a command you can run to automatically keep these `dist` versions up to date. `yarn watch`, I think.

Well that'll show me assuming repos other people create are as jank as my own. In that case I can add the other packages changes to this PR

nabbydude commented on site/examples/js/inlines.jsx:
```diff
@@ -370,8 +367,8 @@ const ToggleEditableButtonButton = () => {
-      onMouseDown={event => {
-        event.preventDefault()
+      onPointerDown={event => event.preventDefault()}
+      onClick={() => {
```
these were changed by `yarn tsc:examples`, they already existed in the tsx files and weren't synced

dylans commented:
@12joan anything else you'd like to see changed with this PR?

12joan commented on packages/slate/src/interfaces/node.ts:
```diff
@@ -322,19 +322,19 @@ export const Node: NodeInterface = {
   },
 
   extractProps(node: Node): NodeProps {
-    if (Element.isAncestor(node)) {
```
I'm thinking there are more reasons to include it than to exclude it, even if it's a wrapper for `!Text.isTextNode`.

nabbydude replied:
Understandable, I'll wait until names are settled and then add it if these names stay

nabbydude commented:
@dylans Can do 🫡
Any thoughts on naming/replacing the existing functions?
- Replacing the existing would be ideal imo but could potentially break some code--any match functions would be fine as is, and most technically-type-unsafe things (like plate calling it with their `At` type, which is either a node or location), would also work just fine as is, but anything relying on it to gracefully handle nullish values for them could throw.
- `Node.isEditor` etc is probably my preferred alternative looking back after the fact and if the above option is off the table and there are no objections I'm going to change it to this

nabbydude commented:
Switched to `Node.isX` since its terse and descriptive, and it doesn't risk any subtle breakages downstream. I'd love to deprecate+remove/rename the classic functions but I'll save that until I (or you, random bystander reading this right now 👁️) add some proper guidance in the docs in favor of these new functions.

Also added `Node.isAncestor` as an alias of `!Node.isText` as requested by @12joan

peerDependencies updated, should be all good to merge now 👍
